questions/250/explanation.md

```diff
@@ -1,62 +1,61 @@
+
+We start by deducing the template arguments for both overloads. For `foo₁`, the argument `1` is used to deduce `T` to be `int`. It's signature is then `foo₁(int, ...)`. However, for overload resolution we ignore the `...` since it doesn't have a matching argument: §[over.match.viable]¶2.3:
+
+> For the purposes of overload resolution, the parameter list is truncated on the right, so that there are exactly *m* parameters.
+
+So we end up with the viable function `foo₁(int)`.
+
+For `foo₂`, `T...` is a function parameter pack. The argument `1`  is used to deduce this as one `int`, and it's signature is `foo₂(int)`.
+
+So, both `foo`s are viable. Which one is best? §[over.match.best]¶2:
+§[over.match.best.general]¶2.5:
+
+> (...) a viable function `F1` is defined to be a better function than another viable function `F2` if (...)
+>
+> — (...)
+>
+> — `F1` and `F2` are function template specializations, and the function template for `F1` is more specialized than the template for `F2` according to the partial ordering rules described in §[temp.func.order] (...)
+
+I won't go into all the details for partial ordering here, but skip to the interesting parts. In short, we transform each specialization by substituting each template parameter with a unique, made-up type and get `foo₁(X, ...)` and `foo₂(Y)`. Then we do deduction from each transformed function to the other, original template.
+
+Again, the ellipsis is ignored since it doesn't have an argument at the call site: §[temp.deduct.partial]¶3:
+
+> The types used to determine the ordering depend on the context in which the partial ordering is done:
+> - In the context of a function call, the types used are those function parameter types for which the function call has arguments.
+
+We now do deduction of  `foo₂(T...)` from `foo₁(X)`, which deduces `T=X`. We then do deduction of `foo₁(T)` from `foo₂(Y)`. It looks like we would get `T=Y`, but §[temp.deduct.type]¶9.2 says:
+
+> During partial ordering, if `A`<sub>*i*</sub> was originally a pack expansion:
+>
+> — (...) if `P`<sub>*i*<sub> is not a pack expansion, template argument deduction fails.
+
+`A`<sub>*i*<sub> in this case is `Y`, which was originally a function parameter pack. `P`<sub>*i*<sub> is `T`, which is not a function parameter pack, so deduction fails.
+
+Since we can deduce `foo₂` from `foo₁`, but not `foo₁` from `foo₂`, `foo₁` is more specialized and a better function. Overload resolution picks it and prints `A`.
+
+### `foo(1,2)`:
+
+Now let's look at the call `foo(1,2)`.
+
+Again we start by deducing the template arguments. For `foo₁`, the first argument `1` is used to deduce `T` to be `int`. Its signature is then `foo₁(int, ...)`.
+
+For `foo₂`, `T...` is a function parameter pack. The arguments `1` and `2` are used to deduce this as two `int`s, and its signature is `foo₂(int, int)`.
+
```

Comment:
```suggestion
For `foo₂`, `T...` is a function parameter pack. The arguments `1` and `2` are used to deduce this as two `int`s, and its signature is `foo₂(int, int)`.

`foo₂` is viable for the same reasons as before, `foo₁` is viable due to §[over.match.viable]¶2.2:

> A candidate function having fewer than `m` parameters is viable only if it has an ellipsis in its parameter list (§[dcl.fct]). For the purposes of overload resolution, any argument for which there is no corresponding parameter is considered to “match the ellipsis” (§[over.ics.ellipsis]).

```

Reply:
Here we can also say that for `2` -> `...` we use an ellipsis conversion sequence instead of an empty standard conversion sequence. But perhaps the explanation is already big enough.

Reply:
Yeah, this is long enough already imo. These are supposed to be instructive, detailed enough to be complete, but not too off-putting for people to actually read.